File: code/ex42.java
// Monotone stack

class Solution {
    public int trap(int[] height) {
        int n = height.length;    if (n == 0)    return 0;
        
        // Build a monotone stack for the highest bar from right to left [i : n - 1]
        Stack<Integer> stack = new Stack<>();
        for (int i = n - 1; i >= 1; --i) {
            if (stack.isEmpty() || height[i] > height[stack.peek()])
                stack.push(i);
        }  
        
        // Search from the left to right for trapped water
        int total = 0, left = height[0];
        for (int i = 1; i < n; i++) {
            if (!stack.isEmpty() && stack.peek() == i)
                stack.pop();
            if (stack.isEmpty())    break;
            
            if (height[i] < left && height[i] < height[stack.peek()]) {
                total += Math.min(left, height[stack.peek()]) - height[i];
            } else if (height[i] >= left) {
                left = height[i];
            }
        }
        
        return total;
    }
}

File: code/ex42.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        
        vector<int> suffix(n);
        for (int i = n - 1; i >= 0; i--)
            suffix[i] = i == n - 1 ? 0 : max(suffix[i + 1], height[i + 1]);
        
        int res = 0, MAX = 0;
        for (int i = 0; i < n; i++) {
            res += max(0, min(MAX, suffix[i]) - height[i]);
            MAX = max(MAX, height[i]);
        }
        
        return res;
    }
};